### ISSUE TITLE
Previewing audio/video files with ffplay?!

### ISSUE THREAD

**Issue by tsujan** [lxqt/pcmanfm-qt#1077]
I just wanted to tell that a nice audio/video preview is possible by using `ffplay` but without adding a dependency.

A while ago, I applied a small patch to pcmanfm-qt here, to have a light preview of selected audio/video files by pressing `F1`. When the selection or directory changes, playing is stopped by killing `ffplay` (`F1` itself can also stop preview). The audio preview has no window.

The code is very elementary and, basically, is nothing but a command-line handled by `QProcess`. That's the reason why I didn't make a PR. However, the result has been pleasant.

Please tell me about your opinion. Negative replies will be happily accepted if based on good enough reasons (actually, I want them).

**Comment by tsujan** [comment]
Problem happens when ffplay hangs and `QProcess` tries to kill it: a 30-second GUI freeze is possible. It happened only once in a month but was annoying. That's the case will all non-detached processes: we can't control the behavior of the external app.

So, this can be closed. A multimedia preview is still possible by using Qt but it isn't worth the amount of extra codes it requires. Instead, the user can simply add a custom action.

**Comment by tsujan** [comment]
Just for the record:

Yesterday and today, I experimented with multimedia preview by using Qt's multimedia libraries. It created unexpected challenges (like disabling screensaver, forcing a single preview at each time, etc.) but they could be handled by using pure Qt methods and adding ~300 lines of code plus 2 new dependencies. It works well but, IMO, clutters pcmanfm-qt's code and contradicts its lightness.

**Comment by tsujan** [comment]
Update (probably, the last one):

Months ago, I made a patch to preview multimedia with pcmanfm-qt and also to see full videos. It was relatively simple and quite maintainable — the job was mostly done by Qt Multimedia.

But, after a while, I saw a serious problem about RAM usage and `valgrind` confirmed it. Here, with a 16-GB RAM, the memory usage of pcmanfm-qt reached 600 MB after previewing many video/audio files for a day. There have been several reports about memory leaks in Qt Multimedia. Although they're closed, `QMediaPlayer` still leaks badly. Also, `valgrind` showed a leak in `gstreamer` (not in the code), caused by the c-tor of `QMediaPlayer`.

Using an attached `QProcess` and calling external players (or a self-made one) wasn't an option because it could cause 30-second freezes under certain circumstances (see `QProcess::~QProcess`). Using a detached `QProcess` was futile too because there would be no control over playing.

Since the memory leak was ridiculous and because I got addicted to previewing soon, I made the dirtiest possible hack to work around the leak.

Although I still don't think that it's a good idea to play audio/video with a file manager — I mean this with an eye to coding principles — I'd share the original patch if there was no leak. As for the hack, it's too extreme to be shared.

All in all, in my opinion, having a light and reasonable multimedia preview with pcmanfm-qt isn't possible yet.
